File: admin-app/src-tauri/src/lint.rs
use crate::{AppError, AppResult, resolve_in_repo};
use serde::Serialize;
use std::collections::HashSet;
use std::path::Path;

#[derive(Serialize, Debug, Clone)]
pub struct LintIssue {
    pub severity: String,
    pub message: String,
    pub file: String,
    pub line: Option<usize>,
}

#[derive(Serialize, Debug, Clone, Default)]
pub struct LintReport {
    pub issues: Vec<LintIssue>,
    pub warnings_count: usize,
    pub errors_count: usize,
}

impl LintReport {
    pub fn has_errors(&self) -> bool {
        self.errors_count > 0
    }
}

/// Lint Markdown posts before a publish: missing frontmatter dates, drafts
/// leaking into publish, broken local links, etc.
pub fn lint_posts(repo: &Path) -> AppResult<LintReport> {
    let pages_root = resolve_in_repo(repo, "content/pages")?;
    let mut report = LintReport::default();

    fn walk(repo: &Path, pages_root: &Path, dir: &Path, report: &mut LintReport) -> AppResult<()> {
        // Each direct child of the pages root is a page directory.
        for entry in std::fs::read_dir(dir).map_err(AppError::from)?.flatten() {
            let page_dir = entry.path();
            if !page_dir.is_dir() {
                continue;
            }
            let posts_dir = page_dir.join("posts");
            if posts_dir.is_dir() {
                let entries = std::fs::read_dir(&posts_dir).map_err(AppError::from)?;
                for post in entries.flatten() {
                    let path = post.path();
                    if !path.is_file() || path.extension().map(|e| e != "md").unwrap_or(true) {
                        continue;
                    }
                    let rel = path.strip_prefix(pages_root).unwrap_or(&path);
                    let rel_str = rel.to_string_lossy().to_string();
                    match lint_one_post(repo, &path, &rel_str) {
                        Ok(mut issues) => report.issues.append(&mut issues),
                        Err(e) => report.issues.push(LintIssue {
                            severity: "error".into(),
                            message: format!("Could not lint: {}", e),
                            file: rel_str,
                            line: None,
                        }),
                    }
                }
            }
            let subpages = page_dir.join("subpages");
            if subpages.is_dir() {
                walk(repo, pages_root, &subpages, report)?;
            }
        }
        Ok(())
    }

    walk(repo, &pages_root, &pages_root, &mut report)?;

    let mut set = HashSet::new();
    report.issues.retain(|i| set.insert(i.file.clone() + &i.message + &i.severity));
    report.errors_count = report.issues.iter().filter(|i| i.severity == "error").count();
    report.warnings_count = report.issues.len() - report.errors_count;
    Ok(report)
}

fn lint_one_post(repo: &Path, file: &Path, rel: &str) -> AppResult<Vec<LintIssue>> {
    let raw = std::fs::read_to_string(file)?;
    let mut issues = Vec::new();
    let (meta, body) = crate::content::parse_post(&raw);

    if meta.status == "published" && (meta.date.is_empty() || meta.date == "null") {
        issues.push(LintIssue {
            severity: "error".into(),
            message: "Published post is missing a date. Set date: YYYY-MM-DD.".into(),
            file: rel.to_string(),
            line: None,
        });
    }
    if meta.title.trim().is_empty() {
        issues.push(LintIssue {
            severity: "error".into(),
            message: "Post is missing a title.".into(),
            file: rel.to_string(),
            line: None,
        });
    }
    if !meta.slug.is_empty() && !meta.slug.chars().all(|c| c.is_ascii_alphanumeric() || c == '-') {
        issues.push(LintIssue {
            severity: "warning".into(),
            message: format!(
                "Slug '{}' may produce an odd URL. Only lowercase letters, digits, hyphens.",
                meta.slug
            ),
            file: rel.to_string(),
            line: None,
        });
    }
    if meta.excerpt.trim().is_empty() && meta.tags.is_empty() {
        issues.push(LintIssue {
            severity: "warning".into(),
            message: "Post has no excerpt and no tags; it may not surface well in search.".into(),
            file: rel.to_string(),
            line: None,
        });
    }

    // Broken local links/images (relative to repo root).
    for (idx, line) in body.lines().enumerate() {
        let mut rest = line;
        while let Some(start) = rest.find("](./") {
            let after = &rest[start + 2..];
            let end = after.find(')').unwrap_or(after.len());
            let target = after[..end].to_string();
            if !target.is_empty() && !repo.join(pages_root_rel(&target)).exists() {
                issues.push(LintIssue {
                    severity: "error".into(),
                    message: format!("Broken internal link/image: {}", target),
                    file: rel.to_string(),
                    line: Some(idx + 1),
                });
            }
            rest = &after[end..];
        }
        let mut i = line;
        while let Some(start) = i.find("![](") {
            let after = &i[start + 4..];
            let end = after.find(')').unwrap_or(after.len());
            let target = &after[..end];
            if !target.is_empty() && !target.starts_with("http") {
                let p = repo.join(target.trim_start_matches('/'));
                if !p.exists() {
                    issues.push(LintIssue {
                        severity: "error".into(),
                        message: format!("Missing image asset: {}", target),
                        file: rel.to_string(),
                        line: Some(idx + 1),
                    });
                }
            }
            i = &after[end..];
        }
    }

    Ok(issues)
}

fn pages_root_rel(target: &str) -> std::path::PathBuf {
    let t = target.trim_start_matches("./");
    std::path::PathBuf::from(t)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn temp_repo(tag: &str) -> std::path::PathBuf {
        std::env::temp_dir().join(format!("studio-lint-{}-{}", tag, std::process::id()))
    }

    #[test]
    fn flags_missing_date_on_published_post() {
        let root = temp_repo("missing-date");
        let dir = root.join("content/pages/demo/posts");
        std::fs::create_dir_all(&dir).unwrap();
        std::fs::write(
            dir.join("p.md"),
            "---\ntitle: Hi\nstatus: published\n---\nHello\n",
        )
        .unwrap();
        let report = lint_posts(&root).unwrap();
        assert!(report.issues.iter().any(|i| i.message.contains("missing a date")));
        assert!(report.has_errors());
        std::fs::remove_dir_all(&root).ok();
    }

    #[test]
    fn clean_post_produces_no_errors() {
        let root = temp_repo("clean");
        let dir = root.join("content/pages/demo/posts");
        std::fs::create_dir_all(&dir).unwrap();
        std::fs::write(
            dir.join("p.md"),
            "---\ntitle: Fine\ndate: 2026-09-01\nstatus: published\nexcerpt: Fine post.\ntags: [A]\n---\nHello\n",
        )
        .unwrap();
        let report = lint_posts(&root).unwrap();
        assert!(!report.has_errors());
        std::fs::remove_dir_all(&root).ok();
    }
}

File: scripts/lint.mjs
import fs from 'node:fs/promises';
import path from 'node:path';
import { PAGES_ROOT } from './content/paths.mjs';
import { parseFrontmatter } from './content/frontmatter.mjs';
import { validatePostMeta, validatePageMeta } from './content/schema.mjs';
import { buildPageHierarchy } from './content/pages.mjs';

const ROOT = process.cwd();

const issues = [];
let sawError = false;

function report(severity, file, line = null, message) {
  issues.push({ severity, file, line, message });
  if (severity === 'error') sawError = true;
}

function isAssetRef(target) {
  const t = target.trimStart().replace(/^\.?\//, '');
  return t.startsWith('assets/');
}

async function targetExists(repo, postAssetsDir, target) {
  const name = target.trimStart().replace(/^\.\//, '');
  if (isAssetRef(name)) {
    const stripped = name.replace(/^assets\//, '');
    const resolved = path.join(postAssetsDir, stripped);
    try {
      await fs.access(resolved);
      return true;
    } catch {
      return false;
    }
  }
  // Refs are resolved against the repo root first, then the post assets dir.
  const asRoot = path.join(repo, name);
  try {
    await fs.access(asRoot);
    return true;
  } catch {
    try {
      await fs.access(path.join(postAssetsDir, name));
      return true;
    } catch {
      return false;
    }
  }
}

async function lintMarkdownEmbeddedLinks(repo, file, body, postAssetsDir) {
  // `](./...` inline links
  for (const [idx, line] of body.split('\n').entries()) {
    let rest = line;
    while (true) {
      const start = rest.indexOf('](./');
      if (start === -1) break;
      const after = rest.slice(start + 2);
      const end = after.indexOf(')');
      const target = end === -1 ? after : after.slice(0, end);
      if (target && !(await targetExists(repo, postAssetsDir, target))) {
        report('error', file, idx + 1, `Broken internal link/image: ${target}`);
      }
      rest = after.slice(end === -1 ? after.length : end);
    }
  }
  let i = body;
  while (true) {
    const start = i.indexOf('![](');
    if (start === -1) break;
    const after = i.slice(start + 4);
    const end = after.indexOf(')');
    const target = end === -1 ? after : after.slice(0, end);
    if (target && !target.startsWith('http') && !(await targetExists(repo, postAssetsDir, target))) {
      report('error', file, lineOf(i, start), `Missing image asset: ${target}`);
    }
    i = after.slice(end === -1 ? after.length : end);
  }
}

function lineOf(text, index) {
  return text.slice(0, index).split('\n').length;
}

async function walkContent(dir, pageSlug) {
  let entries = [];
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const entry of entries) {
    if (!entry.isFile() || !entry.name.endsWith('.md')) continue;
    const file = path.join(dir, entry.name);
    const raw = await fs.readFile(file, 'utf8');
    const { meta, body } = parseFrontmatter(raw);
    const fileRel = path.relative(ROOT, file);

    for (const message of validatePostMeta(meta)) {
      report('error', fileRel, null, message);
    }
    if (meta.page && pageSlug && meta.page !== pageSlug) {
      report('error', fileRel, null, `page '${meta.page}' does not match containing page '${pageSlug}'.`);
    }
    const slug = meta.slug || entry.name.replace(/\.md$/, '');
    await lintMarkdownEmbeddedLinks(ROOT, fileRel, body, path.join(dir, slug, 'assets'));
  }
}

async function walkPage(pageDir) {
  const pageYml = path.join(pageDir, 'page.yml');
  let pageSlug = path.basename(pageDir);
  try {
    const raw = await fs.readFile(pageYml, 'utf8');
    const { meta } = parseFrontmatter(raw);
    pageSlug = meta.slug || pageSlug;
    const pageRel = path.relative(ROOT, pageYml);
    for (const message of validatePageMeta(meta)) {
      report('error', pageRel, null, message);
    }
    if (meta.cover) {
      const cover = String(meta.cover).replace(/^\/+/, '');
      const candidate = cover.includes('/') ? path.join(ROOT, cover) : path.join(pageDir, cover);
      try {
        const stat = await fs.stat(candidate);
        if (!stat.isFile()) throw new Error('not a file');
      } catch {
        report('error', pageRel, null, `Page cover does not exist: ${meta.cover}`);
      }
    }
  } catch (error) {
    if (error?.code === 'ENOENT') {
      report('error', path.relative(ROOT, pageDir), null, 'Page directory is missing page.yml.');
    } else {
      report('error', path.relative(ROOT, pageYml), null, `Could not read page metadata: ${error.message || error}`);
    }
  }

  await walkContent(path.join(pageDir, 'posts'), pageSlug);
  const subpagesDir = path.join(pageDir, 'subpages');
  const children = await fs.readdir(subpagesDir, { withFileTypes: true }).catch(() => []);
  for (const child of children) {
    if (child.isDirectory()) await walkPage(path.join(subpagesDir, child.name));
  }
}

async function main() {
  let pages;
  try {
    pages = await fs.readdir(PAGES_ROOT, { withFileTypes: true });
  } catch {
    report('error', 'content/pages', null, 'content/pages directory is missing.');
    pages = [];
  }

  for (const entry of pages) {
    if (!entry.isDirectory()) continue;
    const pageDir = path.join(PAGES_ROOT, entry.name);
    await walkPage(pageDir);
  }

  try {
    await buildPageHierarchy(PAGES_ROOT, true);
  } catch (error) {
    report('error', 'content/pages', null, String(error?.message || error));
  }

  // sort: errors first, then by file/line
  issues.sort((a, b) => (a.severity === b.severity ? a.file.localeCompare(b.file) || (a.line ?? 0) - (b.line ?? 0) : a.severity === 'error' ? -1 : 1));

  for (const issue of issues) {
    const loc = issue.line ? `${issue.file}:${issue.line}` : issue.file;
    console.log(`${issue.severity.padEnd(7)} ${loc}  ${issue.message}`);
  }
  console.log(`\nlint: ${issues.filter((i) => i.severity === 'error').length} errors, ${issues.filter((i) => i.severity === 'warning').length} warnings`);
  process.exitCode = sawError ? 1 : 0;
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
